### ISSUE TITLE
CMake hygiene

### ISSUE THREAD

**Issue by Orphis** [hrydgard/ppsspp#9013]
I reviewed most CMake files and I concluded that they have a very poor "hygiene" and don't follow current recommendations, standard or use modern commands.

I propose to update the current projects in order to modernize it.
Some recommendations can be read in this blog post: https://rix0r.nl/blog/2015/08/13/cmake-guide/

- No globals, just like in proper C++ code. So include_directories() isn't the way to go, but target_include_directories() should be used instead. Defines should be attached to a target when possible, so no add_definitions() but target_compile_definitions().

- Targets should always include and use files in the current folder or in its subfolders. It should never go up in the hierarchy. It is fine to link with targets wherever they are, preferably without introducing any circular dependency.

- Hermeticity. If you want to include a header from a target, you should link with the target, and it will propagate its include path. In the current state, everything can include anything because of all the "include_directories" we have everywhere. Also, having multiple targets in the same folder won't help since they would both export the same include paths.
 I propose to then update the different libraries:
 - Old: folder/foo.h folder/foo.cpp
 - New: folder/include/folder/foo.h folder/src/foo.cpp. This way, the way to include it is still #include "folder/foo.h" but it won't leak any other header from the internal libraries.

- Fewer CMake options and defines. Most of them can be autodetected using defines from the compiler. A convenience "config.h" could be introduced that translate them to a standard set of defines we need. Features like "use ffmpeg" could stay (though I understand you may want to remove it).

- Platform conditionals should only be used in CMake to switch which libraries you link with, not the files you compile. This will enable a global view of all the files in IDEs. Cpp files will then have the adequate #if around the code that is platform specific.

Dependencies will be more explicit, quality will go up!
I know some of those rules go against the Google C++ guidelines (the #include should match the path from the root of the library), but to be honest, it doesn't work in the real world that doesn't use Blaze that checks the header includes.

Thoughts?

**Comment by Orphis** [comment]
Alright, as for the hermeticity, we could go slowly and start it with the various ext/ libraries first. I understand that core, gpu and hle are too messy right now for that, we (me) could do that later.

**Comment by benderscruffy** [comment]
is this still an issue ??

**Comment by hrydgard** [comment]
Yeah, the CMake is still a mess. It works though, so not critical, as evidenced by this issue being open for 6 years.
